File: leetcode/0001-1000/0649-dota2-senate/Dota2Senate.java
import java.util.ArrayDeque;
import java.util.Deque;

public class Dota2Senate {
  public static void main(String[] args) {
    System.out.println(new Solution().predictPartyVictory("RDRDRDRDR"));
  }
}

class Solution {
  public String predictPartyVictory(String senate) {
    Deque<Integer> qr = new ArrayDeque<>();
    Deque<Integer> qd = new ArrayDeque<>();

    for (int i = 0; i < senate.length(); i++) {
      if (senate.charAt(i) == 'D') {
        qd.offerLast(i);
      } else {
        qr.offerLast(i);
      }
    }

    while (!qr.isEmpty() && !qd.isEmpty()) {
      int r = qr.pollFirst();
      int d = qd.pollFirst();

      if (r < d) {
        qr.offerLast(r + senate.length());
      } else {
        qd.offerLast(d + senate.length());
      }
    }

    return qr.isEmpty() ? "Dire" : "Radiant";
  }
}

File: leetcode/0001-1000/0601-0700/0649-dota2-senate/Dota2Senate.ts
function predictPartyVictory(senate: string): string {
  const qr: number[] = [];
  const qd: number[] = [];

  for (let i = 0; i < senate.length; i++) {
    if (senate[i] === "R") {
      qr.push(i);
    } else {
      qd.push(i);
    }
  }

  while (qr.length > 0 && qd.length > 0) {
    const r = qr.shift();
    const d = qd.shift();

    if (r! < d!) {
      qr.push(r! + senate.length);
    } else {
      qd.push(d! + senate.length);
    }
  }

  return qr.length === 0 ? "Dire" : "Radiant";
};

// console.log(predictPartyVictory("RD"));
// console.log(predictPartyVictory("RDD"));
console.log(predictPartyVictory("RDRDRDRDR"));
